docs/design/2020-06-24-placement-rules-in-sql.md

```diff
@@ -0,0 +1,366 @@
+
+Other DDL on indexes won’t affect placement rules:
+
+* Renaming index
+* Change the visibility of index
+
+### Show DDL jobs
+
+As mentioned earlier, all statements related to placement rules must wait until PD returns. If the execution is interrupted, it should retry until it succeeds or is cancelled, just like other DDL jobs.
+
+PD schedules regions asynchronously after it returns the message. It’s not possible to check the progress of the scheduling for now.
+
+Ongoing and finished statements related to placement rules can also be queried through `ADMIN SHOW DDL`, `ADMIN SHOW DDL JOBS`, or other similar statements.
+
+## View rules
+
+All placement rules should be queried through statements.
+
+### System table
+
+A new system table `information_schema.placement_rules` can be added to view all placement rules. The table contains such columns:
+
+* rule_id
+* database_id and database_name
+* table_id and table_name
+* partition_id and partition_name
+* index_id and index_name
+* label
+* role
+* count
+
+The system table is a virtual table, which doesn’t store data. When querying the table, TiDB queries PD and integrates the result in a table format. That also means the metadata is stored on PD instead of TiKV.
+
+Advantages of building system table include:
+
+* It’s easy for users to filter and aggregate
+* There’s no need to support new grammar, and it’s easier to implement
+
+### Show placement
+
+But there’re a problem here. The system table only contains stored placement rules, and users cannot query the effective rule of one object from it.
+
+For example, table `t` has two partitions `p0` and `p1`, and a placement rule is added on `t`. If the user wants to query the working rule of `p0`. Then he’ll find no placement rule is defined for `p0` through the system table. Based on the rule priorities described before, he must query the placement rule on `t`. This procedure may be annoying.
+
+To simplify this procedure, a SHOW statement should be provided to query the effective rule for one specified object.
+
+The statement is in such a format:
+
+```sql
+SHOW PLACEMENT FOR TABLE {table_name | partition_name};
+SHOW PLACEMENT FOR INDEX index_name FROM {table_name | partition_name};
+```
+
+TiDB will automatically find the effective rule based on the rule priorities.
+
+The output of this statement contains these fields:
+
+* target: The object queried. It can be a table, partition, or index. For table and partition, it is shown in the format `TABLE table_name`; For index, it is shown in the format `INDEX index_name FROM table_name`.
+* placement: The original ALTER statement that affects the placement of this rule. Note that the object in this ALTER statement may be different with `target`. In the example above, if `target` is `p0`, the ALTER statement shown may be `ALTER TABLE t …`.
+
+### Show create table
+
+It’s useful to show rules in `SHOW CREATE TABLE` statement, because users can check the rules easily.
```

Comment:
It is also potentially useful *or* annoying for logical dumpers since a `CREATE TABLE` may fail to restore on a different configuration.

I am not sure what is the best way to handle it: have a session variable that behaves similarly to `set foreign_key_checks=0` in the logical dump? i.e. `set tidb_skip_placement_rule_checks=1`.

@kennytm what do you think?

Reply:
@wwar The conclusion in line 348 is that `SHOW CREATE TABLE` will not be changed.

The benefit of including the rules in `SHOW CREATE TABLE` is to "check the rules easily", but I don't think it is any easier than `SHOW PLACEMENT FOR TABLE`.